Mildwonkey/type func

mildwonkey opened the pull request:
Opening a draft PR to see what people think about this before I put any more time into it! 

The new `type()` function, which is only available for terraform console,
prints out the type of a given value. This is mainly intended for
debugging - it's nice to be able to print out terraform's understanding
of a complex variable.

This was a silly little side project, and I don't know if it's valuable (I like it!) and if the implementation is sound - especially note the change I made if the console is returning a multi-line string. I assume that will break things and needs to be improved. 

[UPDATE: I replaced the original image with the updated current output]
<img width="381" alt="Screen Shot 2020-12-02 at 10 56 04 AM" src="https://user-images.githubusercontent.com/6210214/100896820-0989f000-348d-11eb-8d8e-81d9a2a23385.png">

alisdair commented on command/console.go:
```diff
@@ -123,6 +124,8 @@ func (c *ConsoleCommand) Run(args []string) int {
 		c.showDiagnostics(diags)
 		return 1
 	}
+	scope.AddFunc("type", funcs.TypeFunc)
```
🔍 Neat!

alisdair commented on repl/format.go:
```diff
@@ -43,8 +43,10 @@ func FormatValue(v cty.Value, indent int) string {
+			// If this is a multi-line string, return it as-is
+			if strings.Contains(v.AsString(), "\n") {
+				return v.AsString()
+			}
```
I'm not so sure about this. The goal of the using this formatter was to produce results which would round-trip to the same values. Treating multi-line strings like this just to suit the `type` function breaks this.

I do really like the multi-line output for the `type` result being a unformatted literal type. Could we achieve that some other way? Perhaps using [marks](https://github.com/zclconf/go-cty/blob/master/docs/marks.md), or [a capsule type](https://github.com/zclconf/go-cty/blob/master/docs/types.md#capsule-types)?

mildwonkey replied:
I agree that this isn't the correct solution; when I get back to this PR I may take you up on the rubber duck offer since I'm curious how you envision marks/capsule type working in this case! :)

mildwonkey replied:
🎉 @alisdair I see you already fixed this problem for me and it looks great! 😁 

```
> type(var.object)
<<EOT
object({
    "a": string,
    "b": object({
        "c": bool,
    }),
})

EOT
```

alisdair commented on lang/funcs/conversion.go:
```diff
@@ -85,3 +88,125 @@ func MakeToFunc(wantTy cty.Type) function.Function {
+			attrNames = append(attrNames, name)
+		}
+		sort.Strings(attrNames)
+		b.WriteString("object(map[string]{\n")
```
Displaying the object type seems like one of the most valuable parts of this idea. I think it would be clearer to do so using the same HCL-based format that we use for type constraints, rather than referring to Go-like syntax. For example:

```hcl
object({
  a = string,
  b = object({
    c = bool
  })
})
```

What do you think?

mildwonkey replied:
Ah, there it is! I knew that output looked wrong but just couldn't make that leap, thank you!

alisdair reviewed (commented):
I still really like this idea! Some thoughts inline. I'm happy to discuss/rubber duck this further if that'd be helpful as I'd like to see this ship someday.

nexxai commented:
This would be awesome!  I literally just [wrote a blog post](https://nexxai.dev/how-to-debug-terraform-variable-content-using-this-custom-module/) about a "hack" I use to do this very thing, so a built-in way to access the contents of variables for debugging would be a freakin' godsend!

mildwonkey commented:
I've rebased and addressed @alisdair 's feedback - both his recent work in `repl` and a suggestion re: using marks to skip the HEREDOC syntax greatly embettered this PR.

I've made a handful of improvements to the function, and added tests. I also added documentation, which I'm pretty dissatisfied with - a better example would be examining a datasource's output type, but I'm struggling to come up with a useful example. Suggestions _enthusiastically_ welcome!

alisdair reviewed (commented):
I love it! Some quick notes:

- Reading the docs and fully realizing that `type` is really a function which returns a string makes me think that perhaps the cty marks approach is not as neat a solution as I first thought. I wonder if something like `chomp(type({foo: "bar"}))` might panic. If so, and that seems awkward to unpick, I think the heredoc formatted output is a perfectly fine fallback for now.

- I tried testing the above `chomp` example, and it seems that all the existing functions no longer exist in the console. It looks like `AddFunc` is being called before `Scope.Functions()`, which prevents the other functions from being initialized.

mildwonkey commented:
oh, heck - that call to `Functions()` doesn't happen until some time during the repl session, so I'll need to rethink when and probably how the function gets added. I'll close this for now, thanks for figuring that out for me and sorry - I hadn't paid enough attention to the console tests to notice they didn't call any functions, my bad.

alisdair commented:
One idle thought I had (which may not help or even be possible and yet here I am sharing it 💁‍♂️) is that we might be able to implement `type` as a builtin, like `exit` and `help`. It would sidestep the console-only function issue, and also the concern about folks using `type()` in an expression and having confusing things happen. The builtin could presumably directly call `c.Output`, so no need to worry about marks or whatnot.
